project_forecast_line_bokeh_chart/models/forecast_line.py

```diff
@@ -0,0 +1,43 @@
+# Copyright 2022 Therp BV <https://therp.nl>.
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from odoo import api, models
+
+
+class ForecastLine(models.Model):
+
+    _inherit = "forecast.line"
+
+    @api.model
+    def _read_group_raw(
+        self, domain, fields, groupby, offset=0, limit=None, orderby=False, lazy=True
+    ):
+        """
+        Inherit and add __range key, like in odoo 15
+        https://github.com/odoo/odoo/blob/15.0/odoo/models.py#L2431
+        """
+        result = super()._read_group_raw(
+            domain,
+            fields,
+            groupby,
+            offset=offset,
+            limit=limit,
+            orderby=orderby,
+            lazy=lazy,
+        )
+        dt = [
+            f
+            for f in groupby
+            if self._fields[f.split(":")[0]].type
+            in ("date", "datetime")  # e.g. 'date:month'
+        ]
+        for group in result:
+            if dt:
+                group["__range"] = {}
+            for df in dt:
+                field_name = df.split(":")[0]
+                if group.get(df):
+                    range_from, range_to = group[df][0].split("/")
+                    group["__range"][field_name] = {"from": range_from, "to": range_to}
```

Comment:
So I've looked into why the graph is showing empty, and how it looks like to me is:

- Here, you're filling `from` and `to`
- [Here](https://github.com/OCA/project/blob/15.0/project_forecast_line_bokeh_chart/report/forecast_line_reporting.py#L119), you're using it to retrieve `date`, which is then used as a [dictionary key](https://github.com/OCA/project/blob/15.0/project_forecast_line_bokeh_chart/report/forecast_line_reporting.py#L134). These dictionary keys show in a format `W26 2023` for week; `26 Jun 2023` for day, or `Jun 2023` for month. 
- [Here](https://github.com/OCA/project/blob/15.0/project_forecast_line_bokeh_chart/report/forecast_line_reporting.py#L157) however, you're retrieving the list of dates to iterate over, which you're then using to [get](https://github.com/OCA/project/blob/15.0/project_forecast_line_bokeh_chart/report/forecast_line_reporting.py#L188) from the dictionary. These dates are in format `2023-06-26`.

Thus, the `get` does not use the right key and always retrieves `0`.

Reply:
What would happen if, instead if `from`, you would use the `date:self.granularity` key? You'd get results like:

```
 {'__count': 524,
  '__domain': ['&', ('date', '>=', '2023-06-01'), ('date', '<', '2023-07-01')],
  'amount': 51335.3,
  'date:month': 'juni 2023'}]
```

and then store them in the dict under key `juni 2023`, such as now.

Now, you'll change this [dict](https://github.com/OCA/project/blob/15.0/project_forecast_line_bokeh_chart/report/forecast_line_reporting.py#L132) into an OrderedDict, so that it remembers the order in which keys are added.

Finally, [here](https://github.com/OCA/project/blob/15.0/project_forecast_line_bokeh_chart/report/forecast_line_reporting.py#L188) you won't be picking the data by accessing each date key in turn, but you'll instead just loop all of the keys in the dict, in the correct order.